MLS.Agent/NativeAssemblyLoadHelper.cs

```diff
@@ -1,50 +1,218 @@
+                    yield return ".exe";
+                }
+                else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+                {
+                    yield return ".dylib";
+                }
+                else
+                {
+                    yield return ".so";
+                }
             }
 
-            _resolver = new AssemblyDependencyResolver(path);
+            // linux devs often append version # to libraries I.e mydll.so.5.3.2
+            bool ProbingUseSuffix(string name, string s)
+            {
+                return !(name.Contains(s + ".") || name.EndsWith(s));
+            }
+
+            // If the name looks like a path or a volume name then dont prefix 'lib'
+            bool ProbingUsePrefix(string name)
+            {
+                return name.IndexOf(Path.DirectorySeparatorChar) == -1              // If name has directory information no add no prefix
+                        && name.IndexOf(Path.AltDirectorySeparatorChar) == -1       // Alternate Directory seperator
+                        && name.IndexOf(Path.PathSeparator) == -1                   // Path seperator
+                        && name.IndexOf(Path.VolumeSeparatorChar) == -1;            // Volume seperator
+            }
         }
 
-        public void Handle(string assembly)
+        private IEnumerable<string> ProbingPaths(string probingPath, string name)
         {
-            var currentDomain = AppDomain.CurrentDomain;
-            currentDomain.AssemblyLoad += AssemblyLoaded(assembly);
+            // if name is rooted then it's an absolute path to the dll
+            if (Path.IsPathRooted(name))
+            {
+                if (File.Exists(name))
+                    yield return name;
+            }
+            else
+            {
+                // Check it the dll exists in the probepath root
+                foreach(var pname in ProbingFilenames(name))
+                {
+                    var path = Path.Combine(probingPath, pname);
+                    if (File.Exists(path))
+                        yield return path;
+                }
+                // Grovel tha platform specific subdirectory
+                foreach(var rid in ProbingRids())
+                {
+                    var path = Path.Combine(probingPath, "runtimes", rid, "native");
+
+                    // Check it the dll exists in the rid specific native directory
+                    foreach (var pname in ProbingFilenames(name))
+                    {
+                        var p = Path.Combine(path, pname);
+                        if (File.Exists(p))
+                            yield return p;
+                    }
+                }
+            }
+
+            // Computer valid dotnet-rids for this environment: https://docs.microsoft.com/en-us/dotnet/core/rid-catalog
+            // Where rid is: win, win-x64, win-x86, osx-x64, linux-x64 etc ...
+            IEnumerable<string> ProbingRids()
```

Comment:
Why are we doing this ourselves? This is exactly what `AssemblyDependencyResolver` is meant to solve. It reads the `.deps.json` files and handles any RID fallbacks.

Reply:
We don't have a `.deps.json`. We doing a restore but not a build in order to avoid copying so many files out of the NuGet caches.

Reply:
Maybe we should be generating a `.deps.json` file then. The logic being added here is super fragile and will not work in many cases.

For example, the proposed code won't handle the same scenario I just fixed in Azure Functions - https://github.com/Azure/azure-functions-host/issues/5187. In that case, the native library is located at `runtimes\{RID}\nativeassets\{TFM}`. You won't be able to guess at this path, unless you are planning on knowing how to fall back through TFMs. The current RID support here is also problematic because it won't allow for things like alpine to work.

The more we try to hard-code what the current system does, the more we are going to get it wrong. We should be using as much of the current systems as possible.

Reply:
> Maybe we should be generating a .deps.json file then.

Do you mean by doing a build or do you have another approach in mind?

Reply:
You wouldn't need to do a full build to generate a `.deps.json` file. We have a few levels of places we can get at the current code.

1. Call into the [MSBuild Task directly](https://github.com/dotnet/sdk/blob/d8ef7ef571f035fa10f8e13935aa6afe54316579/src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs#L19-L22).
2. Invoke the [MSBuild Target](https://github.com/dotnet/sdk/blob/be128d0ba81f0971ddbd5bb5a1039dd2c7768edf/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L158-L166)

Maybe @dsplaisted or @nguerrera have other ideas on how to generate a `.deps.json` file without doing a full build.

Another potential route would be to inspect NuGet's `project.assets.json` file and read all of the native asset paths from it. However, we would still need to know how to do RID fallback which the `AssemblyDependencyResolver` solves for us. But that is a solvable problem by reading the shared frameworks `.deps.json` file, which will give you the RID fallback graph for the current machine. (Assuming we are always running on a shared framework in dotnet-try.)

Also - another route is that if you just want to avoid "copying so many files out of the NuGet cache", you can set MSBuild properties to disable that. See https://github.com/dotnet/sdk/commit/1936ebd7307cebd6c5d1d8d699b9844f73b8fb2c.

> The `CopyLocalLockFileAssemblies` property can be set to `false` to revert to
the previous behavior for `dotnet build`.

Reply:
I agree with Eric that you should try to reuse as much of the existing build logic as possible.  I would avoid calling the `GenerateDepsFile` task directly, as a lot of the logic of what flows into that happens in targets files.  Running the `GenerateDepsFile` target is probably the right thing to do.